# nf-core/callingcards: Output

## Introduction

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

In the output directory of a given run of this pipeline, there will be a subdirectory for each sample in the [samplesheet](./usage.md#samplesheet-input) where the
directory name is the sample name (first field in the samplesheet).

For example, if you set the paramter `save_intermediates`, and your samplesheet has two samples and looks like this:

```raw
sample,fastq_1,fastq_2,barcode_details
AAV9_cortex,data/AAV9_1-1_cortex_L004_R1.fastq.gz,,barcode_details.json
AAV9_hindbrain,data/AAV9_1-1_hindbrain_L004_R1.fastq.gz,,barcode_details.json
```

Then the output directory would have the following structure:

```raw
example_results/
├── AAV9_cortex
│   ├── bwa
│   ├── fastqc
│   ├── hops
│   ├── picard
│   ├── processed_reads
│   ├── samtools
│   └── subread
├── AAV9_hindbrain
│   ├── bwa
│   ├── fastqc
│   ├── hops
│   ├── picard
│   ├── processed_reads
│   ├── samtools
│   └── subread
├── genome
│   └── bwa
├── multiqc
│   ├── multiqc_data
│   ├── multiqc_plots
│   └── multiqc_report.html
└── pipeline_info
    ├── execution_report_2023-01-23_11-55-37.html
    ├── execution_timeline_2023-01-23_11-55-37.html
    ├── execution_trace_2023-01-23_11-55-37.txt
    ├── pipeline_dag_2023-01-23_11-55-37.html
    ├── samplesheet.valid.csv
    └── software_versions.yml

```

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

- [Preprocessing](#preprocessing)
    - [UMITools](#umi-tools-extract) - extract Calling Cards specific non-genomic sequence and save the sequence
    for downstream processing
    - [Trimmomatic](#trimmomatic) - trim reads after extracting known Calling Cards specific non-genomic sequence
    - [`Samtools faidx`](#samtools-faidx) - generate a .fai fasta index of the genome if necessary
    - [FastQC](#fastqc) - raw read level QC
    - [Aligner Specific Indexing](#alignment) - generate an index for the selected aligner if necessary
- [Alignment](#alignment) - Align pre-processed reads with one of several aligners
- [Process Alignments](#process-alignments)
    - [Samtools Alignment QC Metrics](#samtools) - bam indexing and extract alignment level QC metrics
    - [picard](#picard) - additional QC metrics
    - [subread featurecounts](#featurecounts) - Provides an overview of
    - [Counting Hops](#pycallingcards) - Iterate over alignments and count qualifiying reads
- [MultiQC](#multiqc) - Collect and display logs and reports
- [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution

## Preprocessing
### UMI-tools extract

<details markdown="1">
<summary>Output files</summary>

- `umitools/`
  - `*.fastq.gz`: The input fastq files with the specified pattern extracted and placed in the
  header. For example, for single-end reads where the `r1_bc_pattern` is
  `NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN`, the result of this step would turn a FASTQ record like this:

  ```raw
    @A00118:503:HNGFYDSX3:4:1101:2844:1000 1:N:0:CCTACCAAAG+TATGTTTATG
    TAGCGTCAATTTTACGCAGACTATCTTTCTGAGGTTAAAGTCTACCCACTTATTGAATTGTATATGTGTGATTCAGTTTTGCCAATGATGCTTTCCAGTCTCTAAAACGGATTTATCCTGGGTAACAAGAGCCTTGCACATTTTGATGACT
    +
    FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFF

  ```

  to this:

  ```raw
    @A00118:503:HNGFYDSX3:4:1101:2844:1000_TAGCGTCAATTTTACGCAGACTATCTTTCTGAGGTTAA 1:N:0:CCTACCAAAG+TATGTTTATG
    AGTCTACCCACTTATTGAATTGTATATGTGTGATTCAGTTTTGCCAATGATGCTTTCCAGTCTCTAAAACGGATTTATCCTGGGTAACAAGAGCCTTGCACATTTTGATGACT
    +
    FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFF
  ```

  Where the barcode immediately follows the first space delimited value in the id line and is appended with an `_`

  - `*.log`: Log file generated by the UMI-tools `extract` command.

</details>

[UMItools extract](https://github.com/CGATOxford/UMI-tools) is used to extract barcode and other CallingCards specific sequences from
single or paired-end reads.

### Trimmomatic

<details markdown="1">
<summary>Output files</summary>

- `trimmomatic/`
  - `*.fastq.gz`: The fastq file, after the [`UMITools extract`](#umi-tools-extract)
  - `*.log`: Log file generated by tyrimmomatic

</details>

[Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) is used to trim reads after extracting known non-genomic
Calling Cards specific sequence

### FastQC

<details markdown="1">
<summary>Output files</summary>

- `fastqc/`
  - `*_fastqc.html`: FastQC report containing quality metrics.
  - `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

</details>

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

![MultiQC - FastQC sequence counts plot](images/mqc_fastqc_counts.png)

![MultiQC - FastQC mean quality scores plot](images/mqc_fastqc_quality.png)

![MultiQC - FastQC adapter content plot](images/mqc_fastqc_adapter.png)

> **NB:** The FastQC plots displayed in the MultiQC report shows reads _after_ preprocessing steps.

### SAMtools

<details markdown="1">
<summary>Output files</summary>

- `<ALIGNER>/`
  - `<SAMPLE>.sorted.bam`: Coordinate sorted alignment file generated by the user-selected aligner
  - `<SAMPLE>.sorted.bam.bai`: the BAI index for the alignment file
- `<ALIGNER>/samtools_stats/`
  - SAMtools `<SAMPLE>.sorted.bam.flagstat`, `<SAMPLE>.sorted.bam.idxstats` and `<SAMPLE>.sorted.bam.stats` files generated from the alignment files.

</details>

The original BAM files generated by the selected alignment algorithm are further processed with [SAMtools](http://samtools.sourceforge.net/) to sort them by coordinate, for indexing, as well as to generate read mapping statistics.

![MultiQC - SAMtools alignment scores plot](images/mqc_samtools_mapped.png)

![MultiQC - SAMtools mapped reads per contig plot](images/mqc_samtools_idxstats.png)

## Alignment


The user may select any one of the following:

- [bwa aln](https://github.com/lh3/bwa)
- [bwa mem2](https://github.com/bwa-mem2/bwa-mem2)
- [bowtie](https://bowtie-bio.sourceforge.net/index.shtml)
- [bowtie2](https://bowtie-bio.sourceforge.net/bowtie2/index.shtml)

If `save_intermediate` is set, and the selected Aligner generates an index,
the index will be saved at the main level of the `outdir` in a directory called
`genome`. In each sample subdirectory, there will also be a directory
named by the name of the selected aligner which will store the aligner's output.

<details markdown="1">
<summary>Output files</summary>

- `genome/`
  - `<ALIGNER>`: this will store the selected aligners index, if `save_intermediate` is set.

- `<ALIGNER>/`
  - `<SAMPLE>.sorted.bam`: Coordinate sorted alignment file generated by the user-selected aligner
  - `<SAMPLE>.sorted.bam.bai`: the BAI index for the alignment file
</details>


## Process Alignments

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

### MultiQC

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
